### ISSUE TITLE
Restarting server using in game command causes WebConsole to be unusable.

### ISSUE THREAD

**Issue by awesomemoder316** [mesacarlos/WebConsole#78]
I found this strange behaviour, where if I start start the game using startup commands, it runs fine, but if restarted using Minecraft's "restart" command, it does not work, and will report as offline. 

WebConsole version: v2.1
SSL enabled: no
Htttps enabled: no

Error: 
[21:38:31 ERROR]: Shutdown due to fatal error
java.net.BindException: Address already in use
	at sun.nio.ch.Net.bind0(Native Method) ~[?:?]
	at sun.nio.ch.Net.bind(Net.java:552) ~[?:?]
	at sun.nio.ch.ServerSocketChannelImpl.netBind(ServerSocketChannelImpl.java:336) ~[?:?]
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:294) ~[?:?]
	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:88) ~[?:?]
	at org.java_websocket.server.WebSocketServer.doSetupSelectorAndServerThread(WebSocketServer.java:533) ~[?:?]
	at org.java_websocket.server.WebSocketServer.run(WebSocketServer.java:341) ~[?:?]
	at es.mesacarlos.webconsole.WebConsole$1.run(WebConsole.java:94) ~[?:?]
	at java.lang.Thread.run(Thread.java:831) [?:?]
[21:38:31 WARN]: Exception in thread "WebSocketSelector-55" java.lang.NullPointerException: Cannot invoke "org.java_websocket.WebSocket.getRemoteSocketAddress()" because "conn" is null
[21:38:31 WARN]: 	at es.mesacarlos.webconsole.websocket.WSServer.onError(WSServer.java:80)
[21:38:31 WARN]: 	at org.java_websocket.server.WebSocketServer.handleFatal(WebSocketServer.java:649)
[21:38:31 WARN]: 	at org.java_websocket.server.WebSocketServer.doSetupSelectorAndServerThread(WebSocketServer.java:542)
[21:38:31 WARN]: 	at org.java_websocket.server.WebSocketServer.run(WebSocketServer.java:341)
[21:38:31 WARN]: 	at es.mesacarlos.webconsole.WebConsole$1.run(WebConsole.java:94)
[21:38:31 WARN]: 	at java.base/java.lang.Thread.run(Thread.java:831)

I realised that if I restarted the server again, it may function. So it only works sometimes after I restart it.

**Comment by mesacarlos** [comment]
Hello,
Thank you for the detailed explanation of the issue and sorry for my late response.
The problem seems to be that the server is not disabling WebConsole properly therefore the port used by the plugin keeps used by the old instance.

I'll investigate for a solution to this issue.

**Comment by mesacarlos** [comment]
Hi,
I was unable to reproduce this issue. What OS and version of Spigot/Paper are you using?

**Comment by awesomemoder316** [comment]
Hi, I'm on Purpur-1141, 1.16.5, java 16, on Oracle Arm linux. I'm wondering here if the aarch64 is causing unexpected behaviour...

**Comment by awesomemoder316** [comment]
Okay so I did a bit more testing, and here's what I found:

Restarting seems to be okay when I add these flags to startup: -XX:+UseZGC -XX:+ZUncommit -XX:ZUncommitDelay=3600 -XX:+ZProactive -XX:+AlwaysPreTouch -XX:+DisableExplicitGC

Reloading still causes problems (but no one should be using reload anyway). 

Without the flags, I find that my WebConsole will still give the error sometimes. 

I greatly apologise for any inconvenience caused, as I'm sure that you took time out of your busy day to investigate this issue.

**Comment by awesomemoder316** [comment]
If you got an email saying that this issue is resolved, please disregard it. I have found the root problem. 

It seems that when one is logged in to the WebConsole panel, and the server is restarting, there would be an error. When logged in, and a restart is issued, a CME would be thrown before the server shuts down. After it starts back up, the plugin will be unable to connect to the panel, as the panel already has a connection and is occupying the port. This is my hypothesis on the issue. 

To replicate: 
1. Launch server normally.
2. Log into WebConsole panel for the server
3. Restart the server while still being logged in.
4. The server should restart, but the WebConsole would not show that the server is up, due to the errors shown in my first message.

It is important to note that if you get off the WebConsole page for the server while it is shutting down and before it starts back up, there would be no errors. But by idling at the WebConsole page when it is starting back up causes error. 

Maybe getting the WebConsole panel to disconnect from the port would do the trick?

**Comment by mesacarlos** [comment]
Hi,
Thank you for your detailed explanation, again! :) (seriously, most people just type in the title of the issue "dont work lol" and nothing helpful else..)
I just did a small change on the logout handler, the built .jar is in my Jenkins:
https://jenkins.mesacarlos.es/job/WebConsole/lastSuccessfulBuild/artifact/target/WebConsole-2.1-jar-with-dependencies.jar

Could you please try with that build if the problem is solved?

**Comment by awesomemoder316** [comment]
Hello, 
No problem for the detailed explanation! I've made some small things myself, and it's really frustrating when people don't give a proper error log/explanation of what's wrong. It's in both our interest to solve this issue, so I don't really see it as a bother. 

I've did some testing (for about 10 restarts) and it seems like the error still persists. I still find it really strange that a restart causes issues, while a stop proceeded by a start gives no problems. (one would think that there should be no difference, since they are supposedly the same?) 

I've also noticed that by stopping and starting the server, the port seems to be closed from the panel side as well? I have no problems if I stay on the disconnected panel page after stopping and starting, with subsequent restarts having no negative consequence. However, I doubt that that's what is causing the disconnect, and that it was just a coniencidence. I believe that it is more likely due to a timeout on the WebConsole panel side causing a disconnect from the port, as I also have no problems if I leave the panel idle disconnected for a few minutes.

TL;DR
1. Logging into WebConsole panel (regardless of connected or not) and restarting will cause the below error. 
2. Idling on the WebConsole panel when disconnected for a few minutes then restarting will not cause the below error. 

If there is anything you need clarifications with, just send a message :)

Attached is the updated error log: 

[22:10:39 ERROR]: [org.java_websocket.server.WebSocketServer] Shutdown due to fatal error
java.net.BindException: Address already in use
        at sun.nio.ch.Net.bind0(Native Method) ~[?:?]
        at sun.nio.ch.Net.bind(Net.java:552) ~[?:?]
        at sun.nio.ch.ServerSocketChannelImpl.netBind(ServerSocketChannelImpl.java:336) ~[?:?]
        at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:294) ~[?:?]
        at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:88) ~[?:?]
        at org.java_websocket.server.WebSocketServer.doSetupSelectorAndServerThread(WebSocketServer.java:533) ~[?:?]
        at org.java_websocket.server.WebSocketServer.run(WebSocketServer.java:341) ~[?:?]
        at es.mesacarlos.webconsole.WebConsole$1.run(WebConsole.java:94) ~[?:?]
        at java.lang.Thread.run(Thread.java:831) [?:?]
[22:10:39 WARN]: Exception in thread "WebSocketSelector-49" java.lang.NullPointerException: Cannot invoke "org.java_websocket.WebSocket.getRemoteSocketAddress()" because "conn" is null
[22:10:39 WARN]:        at es.mesacarlos.webconsole.websocket.WSServer.onError(WSServer.java:80)
[22:10:39 WARN]:        at org.java_websocket.server.WebSocketServer.handleFatal(WebSocketServer.java:649)
[22:10:39 WARN]:        at org.java_websocket.server.WebSocketServer.doSetupSelectorAndServerThread(WebSocketServer.java:542)
[22:10:39 WARN]:        at org.java_websocket.server.WebSocketServer.run(WebSocketServer.java:341)
[22:10:39 WARN]:        at es.mesacarlos.webconsole.WebConsole$1.run(WebConsole.java:94)
[22:10:39 WARN]:        at java.base/java.lang.Thread.run(Thread.java:831)

**Comment by mesacarlos** [comment]
Hi,
I tried logging into the panel with a lot of tabs, and I could not reproduce the issue :(

However I was trying only with Windows 10 and Ubuntu 18 on the server side and Java 8. I'll try installing Oracle Linux and Java 16 to check if the problem is one of these.

**Comment by awesomemoder316** [comment]
On my side I'll try downgrading java, changing the jar, and trying on a x64 machine. I hope it's not one of those annoying aarch64 issues

**Comment by mesacarlos** [comment]
I was finally able to reproduce the issue!
Java 16, Ubuntu 18.04. The problem seems to be with Java 16.
It is weird, the server cannot start because the port is used, presumably by the old instance of the server, but...

![image](https://i.imgur.com/9MG58Nl.png)

The port is not used by any process!! (it is 8080 in my case)
I'm not sure how to solve this problem.. maybe it is a bug of Java itself, not clearing some internal list of used ports..

As a workaround I would suggest using a start.sh file to start your server containing a "while true" that restarts the server on shutdown ([example](https://stackoverflow.com/questions/65587149/how-do-i-make-a-while-true-loop-only-start-the-program-minecraft-server-again-if))

**Comment by awesomemoder316** [comment]
Thank you very much!

**Comment by mesacarlos** [comment]
Hello one more time!
I kept investigating about this problem and I found that there is a timeout for port bindings and a method called setReuseAddr(boolean) in Java ServerSockets that, when set to true, allows to bind the port again to a service ignoring that timeout.

You can download the updated .jar from my Jenkins: https://jenkins.mesacarlos.es/job/WebConsole/56/artifact/target/WebConsole-2.1-jar-with-dependencies.jar

Hope this build fixes this issue!

**Comment by awesomemoder316** [comment]
I'll give it a try later! Thanks for your help!

**Comment by awesomemoder316** [comment]
Yep, the issue seems to be completely solved now! (Just that the plugin will fail to disable when /rl, but as I said, no one should be using /reload command.
